### PR TITLE
8 create send friend request

### LINKED ISSUE #8: Create Send Friend Request
Endpoint: friends/api/v2/add-friend
Method: POST
attributes: 
- Authorization token in Header
- user_email: str
This will be the email of the user to whom the main user will send the friend request to.

### PR DESCRIPTION
**Pull request by animeshrick**
### Overview

> Github Issue:[ #8 ](https://github.com/KoushikMallik-developer/split-it-api-v2/issues/8)

**What was the problem:**

- new endpoint to add friend request

**How was this solved:**

- add a new endpoint to add friend request
- get email from request
- verify the user for existence
- checked receiver is already an old friend
- checked receiver is received a request from sender
- checked sender cant be receiver
- Checked if the reverse friend request exists (receiver has sent to sender)
- add friend request Obj to DB

**Features:**
- One friend can send friend request to other friends

**Services to be impacted:**
- `split-it-api-v2`

**Example Code or Screenshot:**
```
Example Code or Screenshot
```

**_For the PR Creator_**

- [x] Filled out entirely of PR Template.
- [x] Changes meet acceptance criteria defined in ticket.
- [x] Unit Tests Passed. :tada:
- [x] Tested on DEV Env.
- [ ] Tested on QA Env.

### CODE REVIEW

File: friends/services/friends_services.py [hunk 1]
```diff
@@ -86,3 +94,80 @@ def get_all_received_friend_requests(user_id: str) -> Optional[ExportFriendReque
...
+                    "errorMessage": f"{receiver_info} has already sent you a friend request.",
+                }
+
+            # Check if they are already friends
+            if already_a_friend:
+                return {
+                    "successMessage": None,
+                    "errorMessage": f"You are already friends with this {receiver_info}.",
+                }
+
```
Comment (KoushikMallik-developer) [inline friends/services/friends_services.py]: These all should be handled with exception handling.

**Review by KoushikMallik-developer — CHANGES_REQUESTED** [review]

File: friends/services/friends_services.py (hunk 1 above)
Comment (animeshrick) [reply friends/services/friends_services.py]: exceptions are handled.
Please check when user not found error.
"errorMessage": "InternalServerError: [ErrorDetail(string='User does not exists.', code='invalid')]"

File: friends/serializers/friend_serializer.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+from auth_api.export_types.validation_types.validation_result import ValidationResult
+from auth_api.services.helpers import validate_user_email
+from friends.models.friend import Friend
+
+
+class FriendSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Friend
+        fields = "__all__"
```
Comment (KoushikMallik-developer) [inline friends/serializers/friend_serializer.py]: Serializer should be FriendRequestSerializer right?
and the model should be FriendRequest I guess?

File: friends/urls.py [hunk 3]
```diff
@@ -22,4 +23,9 @@
         MyReceivedFriendRequestsView.as_view(),
         name="All-Received-Requests",
     ),
+    path(
+        "add-friend",
+        AddFriend.as_view(),
+        name="All-Received-Requests",
+    ),
```
Comment (KoushikMallik-developer) [inline friends/urls.py]: Why the name is All-Received-Requests?

**Review by KoushikMallik-developer — CHANGES_REQUESTED** [review]

File: friends/admin.py [hunk 4]
```diff
@@ -1,7 +1,33 @@
...
+    list_filter = ("sender", "receiver")
+    search_fields = ("sender__username", "receiver__username")
+
+    def get_queryset(self, request):
+        """
+        Customize the queryset to filter the FriendRequest model.
+        Example: Show only active senders.
+        """
+        queryset = super().get_queryset(request)
+        return queryset.filter(sender__is_active=True)
```
Comment (KoushikMallik-developer) [inline friends/admin.py]: Do we have inactive friend requests too?

File: friends/friend_exceptions/friend_exceptions.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    def __init__(
+        self,
+        receiver_info: str,
+        msg: Optional[str] = None,
+    ):
+        if not msg and receiver_info:
+            self.msg = f"You are already friends with this {receiver_info}."
+        else:
+            super().__init__(msg)
+        logging.error(self.msg)
```
Comment (KoushikMallik-developer) [inline friends/friend_exceptions/friend_exceptions.py]: Make receiver_info optional, and also handle the case where reciever_info is not provided.

File: friends/services/friends_services.py [hunk 6]
```diff
@@ -86,3 +100,68 @@ def get_all_received_friend_requests(user_id: str) -> Optional[ExportFriendReque
...
+            # Check if a friend request already exists
+            if already_sent_request:
+                raise AlreadyFriendRequestSentError()
+
+            # Check if the reverse friend request exists (receiver has sent to sender)
+            if reversed_sent_request:
+                raise ReversedFriendRequestError(receiver_info)
+
+            # Check if they are already friends
+            if already_a_friend:
```
Comment (KoushikMallik-developer) [inline friends/services/friends_services.py]: Can these validations be done in Serializer?

File: friends/views/add_friend.py
Comment (KoushikMallik-developer) [inline friends/views/add_friend.py:1]: Test it for Invalid tokens as well.

**Review by KoushikMallik-developer — CHANGES_REQUESTED** [review]

File: friends/admin.py [hunk 7]
```diff
@@ -1,7 +1,33 @@
...
+    list_filter = ("sender", "receiver")
+    search_fields = ("sender__username", "receiver__username")
+
+    def get_queryset(self, request):
+        """
+        Customize the queryset to filter the FriendRequest model.
+        Example: Show only active senders.
+        """
+        queryset = super().get_queryset(request)
+        return queryset.filter()
```
Comment (KoushikMallik-developer) [inline friends/admin.py]: What is this method actually doing? Can you please explain?

File: friends/services/friends_services.py [hunk 8]
```diff
@@ -86,3 +94,38 @@ def get_all_received_friend_requests(user_id: str) -> Optional[ExportFriendReque
...
+
+        receiver_info: str = (
+            receiver.username.upper() if receiver.username else receiver.email
+        )
+
+        if receiver:
+            data: dict = {
+                "Sender": sender,
+                "Receiver": receiver,
+                "UserEmail": request_data.user_email,
```
Comment (KoushikMallik-developer) [inline friends/services/friends_services.py]: Why is this `UserEmail` needed?

File: friends/serializers/friend_serializer.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+        receiver: User = data.get("Receiver")
+        user_email: str = data.get("UserEmail")
+
+        email_data: dict = {"email": user_email}
+
+        new_friend_request = FriendRequest(sender=sender, receiver=receiver)
+
+        if self.validate(sender=sender, data=email_data):
+            new_friend_request.save()
+            return new_friend_request
```
Comment (KoushikMallik-developer) [inline friends/serializers/friend_serializer.py]: Why are you using email_data when you have sender and receiver already?

File: friends/serializers/friend_serializer.py [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+from friends.models.friend import Friend
+from friends.models.friend_request import FriendRequest
+
+
+class FriendSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = FriendRequest
+        fields = "__all__"
+
+    def validate(self, sender: User, data: Optional[dict] = None) -> Optional[bool]:
```
Comment (KoushikMallik-developer) [inline friends/serializers/friend_serializer.py]: why are you taking sender also in arguments?

**Review by KoushikMallik-developer — CHANGES_REQUESTED** [review]

**Review by KoushikMallik-developer — DISMISSED** [review]

**Review by KoushikMallik-developer — APPROVED** [review]

**Pull request merged**